File: python/src/scmarkeragent/ortho_map.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import pandas as pd

from .config import ORTHO_DIR

SPECIES = ("Human", "Mouse", "Rat")


class OrthoMap:
    def __init__(self, cache=ORTHO_DIR):
        self.cache = cache
        self._hub = {}
        self._pairs = {}
        self._one_to_one = {}

    def _load_hub(self, other):
        if other in self._hub:
            return self._hub[other]
        f = os.path.join(self.cache, f"ortho_Human_to_{other}.csv")
        if not os.path.exists(f):
            raise FileNotFoundError(
                f"ortholog table missing: {f} (run the ortho export; see ortholog.R)"
            )
        d = pd.read_csv(f, dtype=str)
        d = d.iloc[:, :2]
        d.columns = ["hum", "tgt"]
        d = d.dropna()
        d["hum_key"] = d["hum"].str.upper()
        d["tgt_key"] = d["tgt"].str.upper()
        d = d.drop_duplicates().reset_index(drop=True)
        self._hub[other] = d
        return d

    def pairs(self, src, tgt):
        if src == tgt:
            return None
        if (src, tgt) in self._pairs:
            return self._pairs[(src, tgt)]
        if src not in SPECIES or tgt not in SPECIES:
            raise ValueError(f"unsupported species pair: {src}->{tgt}")
        if src == "Human" and tgt in ("Mouse", "Rat"):
            d = self._load_hub(tgt).rename(
                columns={"hum_key": "src_sym", "tgt": "tgt_sym"}
            )
        elif tgt == "Human" and src in ("Mouse", "Rat"):
            d = self._load_hub(src).rename(
                columns={"tgt_key": "src_sym", "hum": "tgt_sym"}
            )
        else:
            a = self.pairs(src, "Human")
            b = self.pairs("Human", tgt)
            a = a.assign(hub_key=a["tgt_sym"].str.upper())
            d = a.merge(
                b, left_on="hub_key", right_on="src_sym", suffixes=("_a", "_b")
            ).rename(columns={"src_sym_a": "src_sym", "tgt_sym_b": "tgt_sym"})
        d = d[["src_sym", "tgt_sym"]].dropna().drop_duplicates().reset_index(drop=True)
        self._pairs[(src, tgt)] = d
        return d

    def one_to_one(self, src, tgt) -> dict[str, str]:
        key = (src, tgt)
        if key in self._one_to_one:
            return self._one_to_one[key]
        if src == tgt:
            self._one_to_one[key] = {}
            return self._one_to_one[key]
        d = self.pairs(src, tgt).copy()
        d["tgt_key"] = d["tgt_sym"].str.upper()
        fwd_ok = d.groupby("src_sym")["tgt_key"].nunique() == 1
        rev_ok = d.groupby("tgt_key")["src_sym"].nunique() == 1
        d = d[d["src_sym"].isin(fwd_ok[fwd_ok].index) & d["tgt_key"].isin(rev_ok[rev_ok].index)]
        out = dict(zip(d["src_sym"], d["tgt_sym"]))
        self._one_to_one[key] = out
        return out


if __name__ == "__main__":
    om = OrthoMap()
    for a, b in [
        ("Human", "Mouse"),
        ("Mouse", "Human"),
        ("Human", "Rat"),
        ("Rat", "Human"),
        ("Mouse", "Rat"),
        ("Rat", "Mouse"),
    ]:
        pr = om.pairs(a, b)
        print(
            f"{a:5s} -> {b:5s} : {len(pr):6d} ortholog symbol pairs; "
            f"e.g. {list(zip(pr['src_sym'][:3], pr['tgt_sym'][:3]))}"
        )

File: R/inst/scmarkeragent/rflow/ortho_map.R
suppressPackageStartupMessages({
  library(data.table)
})
ORTHO_SPECIES <- c("Human", "Mouse", "Rat")
.ortho_hub <- function(other, ortho_dir) {
  f <- file.path(ortho_dir, sprintf("ortho_Human_to_%s.csv", other))
  if (!file.exists(f)) stop("ortholog table missing: ", f, " (run rflow/ortholog_export.R)")
  d <- data.table::fread(f, colClasses = "character")
  d <- d[, 1:2]
  data.table::setnames(d, c("hum", "tgt"))
  d <- d[!is.na(hum) & nzchar(hum) & !is.na(tgt) & nzchar(tgt)]
  d[, hum_key := toupper(hum)][, tgt_key := toupper(tgt)]
  unique(d)
}
ortho_pairs <- function(src, tgt, ortho_dir) {
  sp <- c("Human", "Mouse", "Rat")
  if (src == tgt) {
    return(NULL)
  }
  stopifnot(src %in% sp, tgt %in% sp)
  if (src == "Human" && tgt %in% c("Mouse", "Rat")) {
    d <- .ortho_hub(tgt, ortho_dir)[, .(src_sym = hum_key, tgt_sym = tgt)]
  } else if (tgt == "Human" && src %in% c("Mouse", "Rat")) {
    d <- .ortho_hub(src, ortho_dir)[, .(src_sym = tgt_key, tgt_sym = hum)]
  } else {
    a <- ortho_pairs(src, "Human", ortho_dir)[, .(s = src_sym, h = toupper(tgt_sym))]
    b <- ortho_pairs("Human", tgt, ortho_dir)
    data.table::setnames(b, c("h", "t"))
    d <- merge(a, b, by = "h", allow.cartesian = TRUE)[, .(src_sym = s, tgt_sym = t)]
  }
  unique(d[!is.na(src_sym) & nzchar(src_sym) & !is.na(tgt_sym) & nzchar(tgt_sym)])
}
ortho_one_to_one <- function(src, tgt, ortho_dir) {
  if (identical(src, tgt)) {
    return(setNames(character(0), character(0)))
  }
  d <- copy(ortho_pairs(src, tgt, ortho_dir))
  if (is.null(d) || !nrow(d)) {
    return(setNames(character(0), character(0)))
  }
  d[, tgt_key := toupper(tgt_sym)]
  fwd_n <- d[, .(n = uniqueN(tgt_key)), by = src_sym]
  rev_n <- d[, .(n = uniqueN(src_sym)), by = tgt_key]
  fwd_ok <- fwd_n[n == 1L, src_sym]
  rev_ok <- rev_n[n == 1L, tgt_key]
  d <- d[src_sym %in% fwd_ok & tgt_key %in% rev_ok]
  setNames(as.character(d$tgt_sym), as.character(d$src_sym))
}
